{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import scanpy.external as sce\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load & Explore the dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The **pbmc3k dataset** is a widely used single-cell RNA-seq dataset that contains around 3,000 peripheral blood mononuclear cells (PBMCs) from a healthy donor. This dataset was generated by 10X Genomics and is commonly used as a benchmark in single-cell analysis pipelines. PBMCs are a diverse collection of immune cells including T cells, B cells, natural killer (NK) cells, and monocytes.\n",
    "\n",
    "## Dataset Construction\n",
    "\n",
    "The pbmc3k dataset is provided in the form of an **AnnData** object, a popular data structure in the Scanpy ecosystem designed specifically for single-cell data. Key components of the AnnData object include:\n",
    "\n",
    "- **adata.X**: The main data matrix where each row represents a cell and each column represents a gene.\n",
    "- **adata.obs**: A pandas DataFrame that contains metadata for each cell (e.g., total counts, number of detected genes).\n",
    "- **adata.var**: A pandas DataFrame that holds metadata for each gene (e.g., gene names, whether a gene is mitochondrial).\n",
    "\n",
    "Below, we load and explore the pbmc3k dataset.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Basic information:\n",
      "AnnData object with n_obs × n_vars = 2700 × 32738\n",
      "    var: 'gene_ids'\n",
      "\n",
      "Cell Metadata (obs): \n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [AAACATACAACCAC-1, AAACATTGAGCTAC-1, AAACATTGATCAGC-1, AAACCGTGCTTCCG-1, AAACCGTGTATGCG-1]\n",
      "\n",
      "Gene Metadata (var):\n",
      "                     gene_ids\n",
      "index                        \n",
      "MIR1302-10    ENSG00000243485\n",
      "FAM138A       ENSG00000237613\n",
      "OR4F5         ENSG00000186092\n",
      "RP11-34P13.7  ENSG00000238009\n",
      "RP11-34P13.8  ENSG00000239945\n",
      "\n",
      "Data Matrix Shape:\n",
      "(2700, 32738)\n"
     ]
    }
   ],
   "source": [
    "# Load the pbmc3k dataset\n",
    "adata = sc.datasets.pbmc3k()\n",
    "\n",
    "# print the basic information of the dataset\n",
    "print(\"Basic information:\")\n",
    "print(adata)\n",
    "\n",
    "# View the first few rows of the cell metadata (obs)\n",
    "print(\"\\nCell Metadata (obs): \")\n",
    "print(adata.obs.head())\n",
    "\n",
    "# View the first few rows of the gene metadata (var)\n",
    "print(\"\\nGene Metadata (var):\")\n",
    "print(adata.var.head())\n",
    "\n",
    "# Print the shape of the main data matrix (cells x genes)\n",
    "print(\"\\nData Matrix Shape:\")\n",
    "print(adata.X.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## pbmc3k Dataset Description\n",
    "\n",
    "Here are its key characteristics:\n",
    "\n",
    "- **Number of Cells (`n_obs`):** 2,700  \n",
    "- **Number of Genes (`n_vars`):** 32,738  \n",
    "- **Data Structure:** Stored in an AnnData object, which is a common format for single-cell data in Python.\n",
    "\n",
    "AnnData Structure:\n",
    "\n",
    "1. **`adata.obs` (Cell Metadata):**  \n",
    "   - **Index:** Unique cell barcodes (e.g., `AAACATACAACCAC-1`).  \n",
    "   - **Columns:** Empty in this particular dataset, as no additional cell-level annotations are provided by default.\n",
    "\n",
    "2. **`adata.var` (Gene Metadata):**  \n",
    "   - **Index:** Gene symbols or IDs (e.g., `MIR1302-10`).  \n",
    "   - **Columns:** Typically contains at least one column, here named `\"gene_ids\"`, which stores the corresponding Ensembl gene IDs.\n",
    "\n",
    "3. **`adata.X` (Main Data Matrix):**  \n",
    "   - A matrix of size `n_obs x n_vars` (2,700 × 32,738 in this case).  \n",
    "   - Each row corresponds to a cell, and each column corresponds to a gene.  \n",
    "   - The values are gene expression counts (or sometimes transformed counts).\n",
    "      - **Gene Expression Counts**: Raw counts (also called UMI counts in many single-cell protocols) represent the number of times a given transcript (gene) was detected in each cell.\n",
    "      - **Transformed Counts**: Transformed counts are derived from raw counts through operations such as normalization and log-transformation. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simulate Batch Effects\n",
    "We simulate batch effects by dividing the dataset into two artificial batches.\n",
    "\n",
    "## Why do we simulate Batch Effects?\n",
    "- **Testing Batch Correction Methods:**\n",
    "The pbmc3k dataset originally comes from a single batch, meaning it lacks inherent batch effects. By artificially dividing the data, we create a scenario where batch effects are present. This allows us to test and validate batch integration methods (like Scanorama) in a controlled setting.\n",
    "\n",
    "- **Mimicking Real-World Data:**\n",
    "Many real-world single-cell experiments involve multiple batches. Simulating batches helps us understand how our analysis pipeline would perform on data where batch-to-batch variability is a concern.\n",
    "\n",
    "## Why Only Two Batches? \n",
    "   - **Common Scenario:** Many single-cell experiments involve two batches, especially when comparing different experimental conditions or replicates.\n",
    "   - **Simplicity in Demonstration:** Using two batches keeps the example straightforward while still demonstrating the need for batch integration. More complex scenarios can be explored by creating additional artificial batches if needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   batch\n",
      "index                   \n",
      "AAACATACAACCAC-1  batch1\n",
      "AAACATTGAGCTAC-1  batch1\n",
      "AAACATTGATCAGC-1  batch1\n",
      "AAACCGTGCTTCCG-1  batch1\n",
      "AAACCGTGTATGCG-1  batch1\n",
      "...                  ...\n",
      "TTTCGAACTCTCAT-1  batch2\n",
      "TTTCTACTGAGGCA-1  batch2\n",
      "TTTCTACTTCCTCG-1  batch2\n",
      "TTTGCATGAGAGGC-1  batch2\n",
      "TTTGCATGCCTCAC-1  batch2\n",
      "\n",
      "[2700 rows x 1 columns]\n"
     ]
    }
   ],
   "source": [
    "# The number of cells\n",
    "n_cells = adata.n_obs\n",
    "\n",
    "# Split into two equal halves based on indices\n",
    "adata.obs['batch'] = np.where(np.arange(n_cells) < n_cells / 2, 'batch1', 'batch2')\n",
    "print(adata.obs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this example, we are using a simple, straightforward criterion to split the dataset into two artificial batches:\n",
    "   - `np.arange(n_cells)` generates an array of indices from 0 to `n_cells - 1`.\n",
    "   - The condition `< n_cells / 2` splits this array into two equal halves.\n",
    "   - Cells with indices in the first half are labeled as `'batch1'`, and the rest are labeled as `'batch2'`.\n",
    "\n",
    "## Why This Specific Criterion?\n",
    "   - **Simplicity:** Dividing the dataset by the index is an easy and reproducible way to split the data evenly.\n",
    "   - **Balance:** By splitting the cells equally, we ensure that each batch has a similar number of cells, which is useful for comparing the performance of batch correction methods.\n",
    "   - **Controlled Simulation:** Since the original pbmc3k dataset does not have natural batch effects, this method allows us to simulate a batch effect scenario in a controlled manner.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quality Control \n",
    "\n",
    "In this step, we compute QC metrics including:\n",
    "- Total counts per cell (The sum of all UMIs)\n",
    "- Number of genes detected per cell\n",
    "- Percentage of mitochondrial gene counts\n",
    "\n",
    "We visualize these metrics using plots and filter out considering MAD (median absolute deviations).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   batch  n_genes_by_counts  log1p_n_genes_by_counts  \\\n",
      "index                                                                  \n",
      "AAACATACAACCAC-1  batch1                781                 6.661855   \n",
      "AAACATTGAGCTAC-1  batch1               1352                 7.210080   \n",
      "AAACATTGATCAGC-1  batch1               1131                 7.031741   \n",
      "AAACCGTGCTTCCG-1  batch1                960                 6.867974   \n",
      "AAACCGTGTATGCG-1  batch1                522                 6.259581   \n",
      "\n",
      "                  total_counts  log1p_total_counts  \\\n",
      "index                                                \n",
      "AAACATACAACCAC-1        2421.0            7.792349   \n",
      "AAACATTGAGCTAC-1        4903.0            8.497807   \n",
      "AAACATTGATCAGC-1        3149.0            8.055158   \n",
      "AAACCGTGCTTCCG-1        2639.0            7.878534   \n",
      "AAACCGTGTATGCG-1         981.0            6.889591   \n",
      "\n",
      "                  pct_counts_in_top_50_genes  pct_counts_in_top_100_genes  \\\n",
      "index                                                                       \n",
      "AAACATACAACCAC-1                   47.748864                    63.279637   \n",
      "AAACATTGAGCTAC-1                   45.502753                    61.023863   \n",
      "AAACATTGATCAGC-1                   41.314703                    53.794856   \n",
      "AAACCGTGCTTCCG-1                   39.029936                    52.898825   \n",
      "AAACCGTGTATGCG-1                   44.852192                    55.657492   \n",
      "\n",
      "                  pct_counts_in_top_200_genes  pct_counts_in_top_500_genes  \\\n",
      "index                                                                        \n",
      "AAACATACAACCAC-1                    74.969021                    88.393226   \n",
      "AAACATTGAGCTAC-1                    71.813176                    82.622884   \n",
      "AAACATTGATCAGC-1                    65.449349                    79.961893   \n",
      "AAACCGTGCTTCCG-1                    66.691929                    82.569155   \n",
      "AAACCGTGTATGCG-1                    67.176351                    97.757390   \n",
      "\n",
      "                  total_counts_mt  log1p_total_counts_mt  pct_counts_mt  \\\n",
      "index                                                                     \n",
      "AAACATACAACCAC-1             73.0               4.304065       3.015283   \n",
      "AAACATTGAGCTAC-1            186.0               5.231109       3.793596   \n",
      "AAACATTGATCAGC-1             28.0               3.367296       0.889171   \n",
      "AAACCGTGCTTCCG-1             46.0               3.850147       1.743085   \n",
      "AAACCGTGTATGCG-1             12.0               2.564949       1.223242   \n",
      "\n",
      "                  pct_counts_in_top_20_genes  \n",
      "index                                         \n",
      "AAACATACAACCAC-1                   28.211483  \n",
      "AAACATTGAGCTAC-1                   26.147257  \n",
      "AAACATTGATCAGC-1                   27.881867  \n",
      "AAACCGTGCTTCCG-1                   24.554756  \n",
      "AAACCGTGTATGCG-1                   34.046891  \n"
     ]
    }
   ],
   "source": [
    "# Mark mitochondrial genes (gene names starting with 'MT-' since the dataset is about human cells)\n",
    "adata.var['mt'] = adata.var_names.str.startswith('MT-')\n",
    "\n",
    "# Compute QC metrics for each cell\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], inplace=True)\n",
    "\n",
    "print(adata.obs.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     gene_ids     mt  n_cells_by_counts  mean_counts  \\\n",
      "index                                                                  \n",
      "MIR1302-10    ENSG00000243485  False                  0          0.0   \n",
      "FAM138A       ENSG00000237613  False                  0          0.0   \n",
      "OR4F5         ENSG00000186092  False                  0          0.0   \n",
      "RP11-34P13.7  ENSG00000238009  False                  0          0.0   \n",
      "RP11-34P13.8  ENSG00000239945  False                  0          0.0   \n",
      "...                       ...    ...                ...          ...   \n",
      "AC145205.1    ENSG00000215635  False                  0          0.0   \n",
      "BAGE5         ENSG00000268590  False                  0          0.0   \n",
      "CU459201.1    ENSG00000251180  False                  0          0.0   \n",
      "AC002321.2    ENSG00000215616  False                  0          0.0   \n",
      "AC002321.1    ENSG00000215611  False                  0          0.0   \n",
      "\n",
      "              log1p_mean_counts  pct_dropout_by_counts  total_counts  \\\n",
      "index                                                                  \n",
      "MIR1302-10                  0.0                  100.0           0.0   \n",
      "FAM138A                     0.0                  100.0           0.0   \n",
      "OR4F5                       0.0                  100.0           0.0   \n",
      "RP11-34P13.7                0.0                  100.0           0.0   \n",
      "RP11-34P13.8                0.0                  100.0           0.0   \n",
      "...                         ...                    ...           ...   \n",
      "AC145205.1                  0.0                  100.0           0.0   \n",
      "BAGE5                       0.0                  100.0           0.0   \n",
      "CU459201.1                  0.0                  100.0           0.0   \n",
      "AC002321.2                  0.0                  100.0           0.0   \n",
      "AC002321.1                  0.0                  100.0           0.0   \n",
      "\n",
      "              log1p_total_counts   ribo     hb  \n",
      "index                                           \n",
      "MIR1302-10                   0.0  False  False  \n",
      "FAM138A                      0.0  False  False  \n",
      "OR4F5                        0.0  False  False  \n",
      "RP11-34P13.7                 0.0  False  False  \n",
      "RP11-34P13.8                 0.0  False  False  \n",
      "...                          ...    ...    ...  \n",
      "AC145205.1                   0.0  False  False  \n",
      "BAGE5                        0.0  False  False  \n",
      "CU459201.1                   0.0  False  False  \n",
      "AC002321.2                   0.0  False  False  \n",
      "AC002321.1                   0.0  False  False  \n",
      "\n",
      "[32738 rows x 10 columns]\n"
     ]
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
